File: server/src/plugins/runtime/node/proxy/notificationManager.ts
import { NamespaceManager } from '../../../../rpc/namespaces.js';

import type { RPCClient } from '@camera.ui/rpc';
import type { Notification, NotificationManager, PluginInfo } from '@camera.ui/sdk';
import type { NotificationManagerNamespaces } from '../../../../rpc/namespaces.js';

interface PublishEnvelope {
  pluginId: string;
  pluginName: string;
  notification: Notification;
}

export class NotificationManagerProxy implements NotificationManager {
  #proxy: RPCClient;
  #plugin: PluginInfo;
  #namespaces: NotificationManagerNamespaces;

  constructor(proxy: RPCClient, plugin: PluginInfo) {
    this.#proxy = proxy;
    this.#plugin = plugin;
    this.#namespaces = NamespaceManager.notificationManagerNamespaces();
  }

  async publish(notification: Notification): Promise<void> {
    const envelope: PublishEnvelope = {
      pluginId: this.#plugin.id,
      pluginName: this.#plugin.name,
      notification,
    };
    await this.#proxy.publish(this.#namespaces.notificationsPublishSubject, envelope);
  }
}

File: server/src/plugins/runtime/python/proxy/notification_manager.py
from __future__ import annotations

from typing import TypedDict

from _camera_ui_tools.camera_ui_rpc import RPCClient
from _camera_ui_tools.camera_ui_sdk import Notification, NotificationManager
from plugins.runtime.python.namespaces import NamespaceManager
from plugins.runtime.python.typings import PluginInfo


class _PublishEnvelope(TypedDict):
    pluginId: str
    pluginName: str
    notification: Notification


class NotificationManagerProxy(NotificationManager):
    def __init__(self, proxy: RPCClient, plugin: PluginInfo) -> None:
        self.__proxy = proxy
        self.__plugin = plugin
        self.__namespaces = NamespaceManager.notification_manager_namespaces()

    async def publish(self, notification: Notification) -> None:
        if not notification.get("title"):
            raise ValueError("notification.title is required")
        envelope: _PublishEnvelope = {
            "pluginId": self.__plugin["id"],
            "pluginName": self.__plugin["name"],
            "notification": notification,
        }
        await self.__proxy.publish(self.__namespaces.notifications_publish_subject, envelope)
